File: daemon/update.py
"""Once-a-day check of PyPI for a newer release. The daemon does the (network) check and writes
~/.browser-daemon/update.json; the CLI only reads that file, so commands stay fast and offline-safe.
Opt out with BROWSER_NO_UPDATE_CHECK=1."""
import json
import logging
import os
import threading
import time
import urllib.request
from pathlib import Path

logger = logging.getLogger(__name__)
PACKAGE = "browser-automation-cli"
CACHE = Path.home() / ".browser-daemon" / "update.json"
INTERVAL = 24 * 3600


def current_version() -> str:
    try:
        from importlib.metadata import version
        return version(PACKAGE)
    except Exception:
        return "0.0.0"


def _key(v: str):
    return tuple(int(p) if p.isdigit() else 0 for p in v.split("."))


def is_newer(latest: str, current: str) -> bool:
    return _key(latest) > _key(current)


def fetch_latest(timeout: float = 5.0) -> str:
    with urllib.request.urlopen(f"https://pypi.org/pypi/{PACKAGE}/json", timeout=timeout) as r:
        return json.load(r)["info"]["version"]


def check_now() -> dict:
    info = {"checked_at": time.time(), "current": current_version()}
    try:
        info["latest"] = fetch_latest()
    except Exception as e:
        info["error"] = str(e)
    try:
        CACHE.parent.mkdir(parents=True, exist_ok=True)
        CACHE.write_text(json.dumps(info))
    except Exception:
        pass
    return info


def read_cache() -> dict:
    try:
        return json.loads(CACHE.read_text())
    except Exception:
        return {}


def notice() -> str:
    """One-line hint for the CLI to print on stderr, or '' if up to date / unknown / opted out."""
    if os.environ.get("BROWSER_NO_UPDATE_CHECK"):
        return ""
    c = read_cache()
    latest, cur = c.get("latest"), current_version()
    if latest and is_newer(latest, cur):
        return f"browser-automation-cli {latest} is available (you have {cur}): uv tool upgrade {PACKAGE}  (set BROWSER_NO_UPDATE_CHECK=1 to silence)"
    return ""


def start_background_checks():
    """Daemon side: check at start (if the cache is older than a day) and then daily."""
    if os.environ.get("BROWSER_NO_UPDATE_CHECK"):
        return

    def loop():
        while True:
            c = read_cache()
            if time.time() - c.get("checked_at", 0) > INTERVAL:
                info = check_now()
                if info.get("latest") and is_newer(info["latest"], info["current"]):
                    logger.info(f"Update available: {PACKAGE} {info['latest']} (running {info['current']})")
            time.sleep(3600)

    threading.Thread(target=loop, daemon=True, name="update-check").start()

File: rust/src/update.rs
//! Daily PyPI check (daemon side) -> ~/.browser-daemon/update.json; the CLI only reads that file.
use serde_json::{json, Value};
use std::path::PathBuf;

pub const PACKAGE: &str = "browser-automation-cli";
pub const CURRENT: &str = env!("CARGO_PKG_VERSION");

pub fn cache_path() -> PathBuf { PathBuf::from(std::env::var("HOME").unwrap_or_default()).join(".browser-daemon").join("update.json") }

pub fn key(v: &str) -> Vec<u64> { v.split(|c| c == '.' || c == '-').map(|x| x.parse::<u64>().unwrap_or(0)).collect() }
fn norm(v: &str) -> String { v.replace("-alpha.", "a").replace("-beta.", "b").replace("-rc.", "rc") }
/// PEP 440-ish: "0.4.0a1" < "0.4.0"; "0.4.0-alpha.1" is the cargo spelling of the same.
pub fn is_newer(latest: &str, current: &str) -> bool {
    let (l, c) = (norm(latest), norm(current));
    let strip = |s: &str| -> (Vec<u64>, bool) { let pre = s.contains('a') || s.contains('b') || s.contains("rc"); let base: String = s.chars().take_while(|ch| ch.is_ascii_digit() || *ch == '.').collect(); (key(&base), pre) };
    let (lk, lpre) = strip(&l); let (ck, cpre) = strip(&c);
    if lk != ck { return lk > ck; }
    !lpre && cpre
}

pub fn fetch_latest() -> Result<String, String> {
    let r = ureq::get(&format!("https://pypi.org/pypi/{PACKAGE}/json")).timeout(std::time::Duration::from_secs(5)).call().map_err(|e| e.to_string())?;
    let v: Value = serde_json::from_str(&r.into_string().map_err(|e| e.to_string())?).map_err(|e| e.to_string())?;
    v["info"]["version"].as_str().map(String::from).ok_or("no version in PyPI response".into())
}

pub fn check_now() -> Value {
    let mut info = json!({"checked_at": std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).map(|d| d.as_secs_f64()).unwrap_or(0.0), "current": CURRENT, "client": "rust"});
    match fetch_latest() { Ok(l) => info["latest"] = json!(l), Err(e) => info["error"] = json!(e) }
    let p = cache_path();
    if let Some(d) = p.parent() { let _ = std::fs::create_dir_all(d); }
    let _ = std::fs::write(&p, info.to_string());
    info
}

pub fn read_cache() -> Value { std::fs::read_to_string(cache_path()).ok().and_then(|t| serde_json::from_str(&t).ok()).unwrap_or(Value::Null) }

/// One-line stderr hint when the cache shows a newer release. Silent if opted out / unknown.
pub fn notice() {
    if std::env::var("BROWSER_NO_UPDATE_CHECK").is_ok() { return; }
    let c = read_cache();
    if let Some(latest) = c["latest"].as_str() {
        if is_newer(latest, CURRENT) {
            eprintln!("{PACKAGE} {latest} is available (you have {CURRENT}): uv tool upgrade {PACKAGE}  (BROWSER_NO_UPDATE_CHECK=1 to silence)");
        }
    }
}

/// Daemon side: check if the cache is older than a day, then daily.
pub fn start_background_checks() {
    if std::env::var("BROWSER_NO_UPDATE_CHECK").is_ok() { return; }
    std::thread::spawn(|| loop {
        let c = read_cache();
        let age = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).map(|d| d.as_secs_f64()).unwrap_or(0.0) - c["checked_at"].as_f64().unwrap_or(0.0);
        if age > 86400.0 { let i = check_now(); if let Some(l) = i["latest"].as_str() { if is_newer(l, CURRENT) { eprintln!("[daemon] update available: {PACKAGE} {l} (running {CURRENT})"); } } }
        std::thread::sleep(std::time::Duration::from_secs(3600));
    });
}
